cpp/tests/shared_fcntl_mutex_test.cc

```diff
@@ -0,0 +1,516 @@
+    // parent process. do pipe end cleanup
+    comms[c].close_unused_ends(PipeComms::PARENT);
+  }
+
+  return Status::OK();
+}
+
+// Issues exit instructions to the child process
+// and waits for it to shutdown
+template <std::size_t CHILDREN>
+arrow::Status ShutdownChildren(std::array<PipeComms, CHILDREN>& comms,
+                               std::array<pid_t, CHILDREN>& child_pid) {
+  for (std::size_t c = 0; c < NCHILDREN; ++c) {
+    int status;
+    ARROW_RETURN_NOT_OK(comms[c].send("exit", PipeComms::PARENT));
+    ARROW_RETURN_NOT_OK(comms[c].expect("ok", PipeComms::PARENT));
+
+    if (auto w_pid = waitpid(child_pid[c], &status, 0); w_pid == -1) {
+      return Status::IOError("Waiting for child ", c, " failed");
+    } else if (status != EXIT_SUCCESS) {
+      return Status::IOError("Child ", c, " failed with exit code", status);
+    }
+  }
+
+  return Status::OK();
+}
+
+static constexpr auto PARENT_CONTEXT = PipeComms::PARENT;
+
+// Test that acquiring read locks increases the number of fcntl readers
+TEST(SharedFcntlMutexTest, FcntlReaders) {
+  std::array<PipeComms, NCHILDREN> comms;
+  std::array<pid_t, NCHILDREN> child_pid;
+
+  ASSERT_OK(SpawnChildren(comms, child_pid));
+  constexpr std::size_t NREADLOCKS = 3;
+
+  // Base case for no locks
+  ASSERT_OK(comms[0].send(kQueryFcntlReaders, PARENT_CONTEXT));
+  ASSERT_OK(comms[0].expect("ok 0", PARENT_CONTEXT));
+
+  for (std::size_t i = 0; i < NREADLOCKS; ++i) {
+    auto thread = std::to_string(i);
+
+    // Acquire a read lock in the first child
+    ASSERT_OK(comms[0].send(kReadLock + " thread "s + thread, PARENT_CONTEXT));
+    ASSERT_OK(comms[0].expect("ok", PARENT_CONTEXT));
+
+    // Confirm number of readers
+    ASSERT_OK(comms[0].send(kQueryFcntlReaders + " thread "s + std::to_string(i),
+                            PARENT_CONTEXT));
+    ASSERT_OK(comms[0].expect("ok " + std::to_string(i + 1), PARENT_CONTEXT));
+  }
+
+  for (std::size_t i = 0; i < NREADLOCKS; ++i) {
+    auto thread = std::to_string(i);
+    // Release a read lock in the first child
+    ASSERT_OK(comms[0].send(kReadUnlock + " thread "s + thread, PARENT_CONTEXT));
+    ASSERT_OK(comms[0].expect("ok", PARENT_CONTEXT));
+
+    // Confirm number of readers
+    ASSERT_OK(comms[0].send(kQueryFcntlReaders + " thread "s + thread, PARENT_CONTEXT));
+    ASSERT_OK(
+        comms[0].expect("ok " + std::to_string(NREADLOCKS - i - 1), PARENT_CONTEXT));
+  }
+
+  ASSERT_OK(ShutdownChildren(comms, child_pid));
+}
+
+// Test that IPC locking and lock observation behaves as expected
+TEST(SharedFcntlMutexTest, InterProcessObservability) {
+  std::array<PipeComms, NCHILDREN> comms;
+  std::array<pid_t, NCHILDREN> child_pid;
+  ASSERT_OK(SpawnChildren(comms, child_pid));
+
+  // Acquire a write lock in the first process
+  ASSERT_OK(comms[0].send(kWriteLock + " thread 1"s, PARENT_CONTEXT));
+  ASSERT_OK(comms[0].expect("ok", PARENT_CONTEXT));
+
+  // Failed to acquire a write lock in the same process and thread
```

Comment:
Is this something you want your locking API to guarantee will fail? It's currently undefined behaviour (because it's undefined behaviour for std::shared_mutex).